### PR TITLE
update go-coap to optimized version

### PR DESCRIPTION
**Pull request by jkralik**
replace github.com/patrickmn/go-cache package

### CODE REVIEW

File: cloud2cloud-connector/service/deviceSubscription.go [hunk 1]
```diff
@@ -40,9 +41,9 @@ func (s *SubscriptionManager) SubscribeToDevice(ctx context.Context, deviceID st
 		linkedCloud:   linkedCloud,
 		subscription:  sub,
 	}
-	err = s.cache.Add(correlationID, data, cache.DefaultExpiration)
-	if err != nil {
-		return fmt.Errorf("cannot cache subscription for device subscriptions: %w", err)
+	_, loaded := s.cache.LoadOrStore(correlationID, cache.NewElement(data, time.Now().Add(time.Minute*10), nil))
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/deviceSubscription.go]: Maybe some constant for `time.Minute*10`, the value is used in multiple places when inserting into the cache.

File: cloud2cloud-connector/service/deviceSubscription.go [hunk 2]
```diff
@@ -40,9 +41,9 @@ func (s *SubscriptionManager) SubscribeToDevice(ctx context.Context, deviceID st
...
+	if loaded {
+		return fmt.Errorf("cannot cache subscription for device subscriptions: subscription with %v already exist", correlationID)
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/deviceSubscription.go]: typo: exist -> exists

File: cloud2cloud-connector/service/devicesSubscription.go [hunk 3]
```diff
@@ -39,9 +40,9 @@ func (s *SubscriptionManager) SubscribeToDevices(ctx context.Context, linkedAcco
 		linkedCloud:   linkedCloud,
 		subscription:  sub,
 	}
-	err = s.cache.Add(correlationID, data, cache.DefaultExpiration)
-	if err != nil {
-		return fmt.Errorf("cannot cache subscription for start subscriptions: %w", err)
+	_, loaded := s.cache.LoadOrStore(correlationID, cache.NewElement(data, time.Now().Add(time.Minute*10), nil))
+	if loaded {
+		return fmt.Errorf("cannot cache subscription for start subscriptions: subscription with %v already exist", correlationID)
 	}
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/devicesSubscription.go:46]: exist -> exists

File: cloud2cloud-connector/service/devicesSubscription.go [hunk 4]
```diff
@@ -39,9 +40,9 @@ func (s *SubscriptionManager) SubscribeToDevices(ctx context.Context, linkedAcco
 		linkedCloud:   linkedCloud,
 		subscription:  sub,
 	}
-	err = s.cache.Add(correlationID, data, cache.DefaultExpiration)
-	if err != nil {
-		return fmt.Errorf("cannot cache subscription for start subscriptions: %w", err)
+	_, loaded := s.cache.LoadOrStore(correlationID, cache.NewElement(data, time.Now().Add(time.Minute*10), nil))
+	if loaded {
+		return fmt.Errorf("cannot cache subscription for start subscriptions: subscription with %v already exist", correlationID)
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/devicesSubscription.go]: exist -> exists

File: cloud2cloud-connector/service/requestHandler.go [hunk 5]
```diff
@@ -50,12 +51,19 @@ func NewRequestHandler(
...
-		ownerClaim:     ownerClaim,
-		provider:       provider,
-		subManager:     subManager,
-		store:          store,
-		provisionCache: cache.New(5*time.Minute, 10*time.Minute),
+		ownerClaim: ownerClaim,
+		provider:   provider,
+		subManager: subManager,
+		store:      store,
+		//provisionCache: cache.New(5*time.Minute, 10*time.Minute),
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/requestHandler.go]: remove me

File: cloud2cloud-connector/service/subscriptions.go [hunk 6]
```diff
@@ -61,15 +62,23 @@ func NewSubscriptionManager(
...
 	return &SubscriptionManager{
 		eventsURL:           EventsURL,
 		store:               store,
 		raClient:            raClient,
 		isClient:            isClient,
 		devicesSubscription: devicesSubscription,
-		cache:               cache.New(time.Minute*10, time.Minute*5),
-		provider:            provider,
-		triggerTask:         triggerTask,
+		// cache:               cache.New(time.Minute*10, time.Minute*5),
```
Comment (Danielius1922) [inline cloud2cloud-connector/service/subscriptions.go]: remove me

File: cloud2cloud-connector/service/deviceSubscription.go (hunk 1 above)
Comment (jkralik) [reply cloud2cloud-connector/service/deviceSubscription.go]: done

File: cloud2cloud-connector/service/deviceSubscription.go (hunk 2 above)
Comment (jkralik) [reply cloud2cloud-connector/service/deviceSubscription.go]: fixed

File: cloud2cloud-connector/service/devicesSubscription.go (hunk 4 above)
Comment (jkralik) [reply cloud2cloud-connector/service/devicesSubscription.go]: fixed

File: cloud2cloud-connector/service/requestHandler.go (hunk 5 above)
Comment (jkralik) [reply cloud2cloud-connector/service/requestHandler.go]: fixed

File: cloud2cloud-connector/service/subscriptions.go (hunk 6 above)
Comment (jkralik) [reply cloud2cloud-connector/service/subscriptions.go]: fixed

File: test/oauth-server/service/httpApi.go [hunk 7]
```diff
@@ -71,15 +73,25 @@ func NewRequestHandler(config *Config, idTokenKey *rsa.PrivateKey, accessTokenKe
 	if err != nil {
 		return nil, fmt.Errorf("cannot create jwk for idToken: %w", err)
 	}
+	authSession := cache.NewCache()
+	authRestriction := cache.NewCache()
+	refreshRestriction := cache.NewCache()
+	add := periodic.New(ctx.Done(), time.Second*5)
+	add(func(now time.Time) bool {
+		authSession.CheckExpirations(now)
+		authRestriction.CheckExpirations(now)
```
Comment (Danielius1922) [inline test/oauth-server/service/httpApi.go:82]: Shouldn't also `refreshRestriction` be here?

Comment (jkralik) [reply test/oauth-server/service/httpApi.go:82]: fixed

**Review by Danielius1922 — APPROVED** [review]

**Review by ondrejtomcik — APPROVED** [review]

**Pull request merged**